### PR TITLE
Improve the downsampling algorithm

### PR DESCRIPTION
**Pull request by KYovchevski**
While preparing the talk for UU, I noticed that the quality of images made with our downsampler is very low compared to other downsamplers. I took the time to research why that is, it turned out that we aren't taking nearly enough samples when sampling down. For example, sampling from 2048x2048 down to 512x512, we would always take a 6x6 kernel to sample, while other samplers would take 12x12, and adapt that number further depending on the ratio between the source and target dimensions.

I also took inspiration from how other downsamplers handle working with large numbers of samples by caching some of the math for reuse.

The result is a new implementation which preserves image quality much better, but is about twice slower. The performance can probably be improved by splitting the ISPC kernel into two - one for 3 channels and one for 4 channels, and doing the branch in rust instead of relying on function pointers in ISPC. We might be able to squeeze out more performance with cache optimizations, but it'd need further looking into.

The old implementation is kept in both ISPC and Rust, and be invoked using `downsample_fast`.

### CODE REVIEW

File: src/lib.rs [hunk 1]
```diff
@@ -63,3 +70,114 @@ pub fn downsample(src: &Image, target_width: u32, target_height: u32) -> Vec<u8>
...
+
+    unsafe {
+        ispc::downsample_ispc::calculate_weight_variables(src, target, variables.as_mut_ptr());
+    };
+
+    let image_scale = src as f32 / target as f32;
+
+    let mut res = Vec::with_capacity(target as usize);
+
+    let mut reuse_heap = HashMap::<_, Rc<Vec<f32>>>::with_capacity(target as usize / 2);
```
Comment (MarijnS95) [inline src/lib.rs:96]: It this something we should `static`ally cache (https://crates.io/crates/once_cell) so that subsequent `downsample()` calls benefit from it? And/or put it in the public signature behind an opaque function so that the caller is in control over how to share - _and when to destroy to free up memory_ - the cache?

File: .gitignore [hunk 2]
```diff
@@ -1,5 +1,6 @@
 # Generated by Cargo
 # will have compiled files and executables
+/.vscode/
 /target/
```
Comment (MarijnS95) [inline .gitignore:4]: Nit: This comment doesn't really apply to `/.vscode/` :grimacing:

File: src/ispc/kernels/lanczos3.ispc [hunk 3]
```diff
@@ -222,9 +220,65 @@ void clean_and_write_4_channels(varying float<4> color, varying uint64 write_add
 }
 
 /// scratch_space must be at least src_height * target_width pixels big
-export void resample_with_cache(uniform uint src_width, uniform uint src_height, uniform uint target_width, uniform uint target_height, uniform uint8 num_channels,
+export void resample_with_cache_3(uniform uint src_width, uniform uint src_height, uniform uint target_width, uniform uint target_height,
+    uniform const Cache * uniform cache, uniform uint8 scratch_space[], uniform const uint8 src_data[], uniform uint8 out_data[]) {
+    // TODO[#Kamen]: Ideally, we should split this function into two versions depending channel count, and branch only once in Rust than twice per sample.
```
Comment (MarijnS95) [inline src/ispc/kernels/lanczos3.ispc:225]: Looks like you did that now.

Anyway, is this even faster than the:

```
void resample_with_cache(uniform const uint num_channels, ...) {}

void resample_with_cache_3(...) {
   resample_with_cache(3, ...);
}

void resample_with_cache_4(...) {
   resample_with_cache(4, ...);
}
```

And instead have to duplicate _the entire body_ of `resample_with_cache()`? Implying ISPC doesn't do constant-folding? What was the perf difference?

File: src/ispc/kernels/lanczos3.ispc [hunk 4]
```diff
@@ -113,3 +113,163 @@ export void resample(uniform uint width, uniform uint height, uniform uint strid
             out_data[(x + y * target_width) * num_channels + i] = s[i];
     }
 }
+
+struct WeightVariables {
+    float src_center;
+    float src_start;
+    float src_end;
+};
```
[the remark this answers is not in the document]
Comment (MarijnS95) [reply src/ispc/kernels/lanczos3.ispc:121]: I had imagined this to be relevant for the sampling as it is likely vectorized over pixels, but this struct isn't used there at all (bit hard to follow the naming of all these types and members) - then this comment is moot!

File: src/ispc/kernels/lanczos3.ispc (hunk 3 above)
Comment (KYovchevski) [reply src/ispc/kernels/lanczos3.ispc:225]: No, when we did the testing, the results remained the same. Duplicating the function with just changing the number of channels we read from/write to does result in a ~20% performance gain however.

**Comment by KYovchevski** [comment]
I did some testing with what affects the performance for the test case we have (3 channels, 2048x2048 -> 512x512), and have made some interesting observations.

 - The branching when reading/writing to the pixel buffers does not seem to cause as significant of a performance drop as first thought. Instead, the drop was caused by the memory reinterpretation that was in `sample_3_channels` and `clean_and_write_3_channels`, where a `float<4>` was interpreted as a `float<3>` to use a single write. Skipping this reinterpretation and giving the 3-channel and 4-channel versions different signatures causes a significant speed up. This however means we need to have both a `float<3>` and a `float<4>` in the function that we can write to in the branch.
 - Making `resample_with_cache` `inline` causes a performance decrease of about 10% for our test case. This is including the usage of ISPC's `assume` hints, which would make sure that the branches are removed.

File: benches/basic.rs [hunk 5]
```diff
@@ -25,19 +27,18 @@ pub fn ispc_downsampler(c: &mut Criterion) {
...
             .data
             .chunks(3)
             .map(|v| RGB::new(v[0], v[1], v[2]))
             .collect::<Vec<_>>();
 
-        let mut dst = vec![RGB::new(0, 0, 0); target_width * target_height];
-
         c.bench_function("Downsample `square_test.png` using resize", |b| {
             b.iter(|| {
+                let mut dst = vec![RGB::new(0, 0, 0); target_width * target_height];
```
Comment (MarijnS95) [inline benches/basic.rs:41]: Don't really want to benchmark allocation as part of the resize test? Maybe not even instantiating the resizer (which we might want to benchmark separately as it iirc pre-computes weight tables?).

File: src/ispc/kernels/lanczos3.ispc [hunk 6]
```diff
@@ -18,97 +18,190 @@ static inline float sinc(float x)
...
+    if (t < filter_scale)
+        return clean(sinc(t) * sinc(t / filter_scale));
     else
         return 0.0f;
 }
 
-struct Image {
-    uniform const uint8* data;
-    uniform int<2> size;
+struct WeightVariables {
```
Comment (MarijnS95) [inline src/ispc/kernels/lanczos3.ispc:31]: Do we have a more descriptive name than "variables"? This has to do with image size/resolution, right?

File: src/ispc/kernels/lanczos3.ispc [hunk 7]
```diff
@@ -18,97 +18,190 @@ static inline float sinc(float x)
...
+
+        out_variables[pixel] = vars;
+    }
 }
 
-static inline float frac(float f) {
-    float absf = abs(f);
-    return absf - floor(absf);
+export void calculate_weights(uniform float image_scale, uniform float filter_scale, uniform const WeightVariables * uniform vars, uniform float * uniform weights) {
+
```
Comment (MarijnS95) [inline src/ispc/kernels/lanczos3.ispc:59]:
```suggestion
```

File: src/ispc/kernels/lanczos3.ispc [hunk 8]
```diff
@@ -18,97 +18,190 @@ static inline float sinc(float x)
...
-            } else if (num_channels == 4) {
-                texel.x = byte_to_float(src_image.data[addr + 0]) * inv_255;
-                texel.y = byte_to_float(src_image.data[addr + 1]) * inv_255;
-                texel.z = byte_to_float(src_image.data[addr + 2]) * inv_255;
-                texel.w = byte_to_float(src_image.data[addr + 3]) * inv_255;
-            }
-
-            col += w * texel;
-            weight += w;
+void clean_and_write_3_channels(varying float<3> color, varying uint64 write_address, uniform uint8* varying dst) {
```
Comment (MarijnS95) [inline src/ispc/kernels/lanczos3.ispc:102]: Here and below: `write_address` and `dst` aren't used separately: can you combine them into one argument and let the caller perform `scratch_space + scratch_write_address`?

Comment (MarijnS95) [inline src/ispc/kernels/lanczos3.ispc:102]: s/clean/clamp?

File: src/ispc/kernels/lanczos3.ispc [hunk 9]
```diff
@@ -18,97 +18,190 @@ static inline float sinc(float x)
...
+    uniform const uint32* weight_counts;
+    uniform const float* const* values;
+};
+
+struct Cache {
+    uniform WeightCollection vertical_weights;
+    uniform WeightCollection horizontal_weights;
+};
+
+uint8<3> sample_3_channels(varying uint64 read_address, const uniform uint8* varying src) {
```
Comment (MarijnS95) [inline src/ispc/kernels/lanczos3.ispc:92]: Combine the arguments?

File: src/ispc/kernels/lanczos3.ispc [hunk 10]
```diff
@@ -18,97 +18,190 @@ static inline float sinc(float x)
...
-    return (float)b;
+struct WeightCollection {
+    uniform const uint32* starts;
+    uniform const uint32* weight_counts;
+    uniform const float* const* values;
+};
+
+struct Cache {
+    uniform WeightCollection vertical_weights;
+    uniform WeightCollection horizontal_weights;
```
Comment (MarijnS95) [inline src/ispc/kernels/lanczos3.ispc:89]: I don't think this counts as a "cache" (since we don't insert nor evict anything) but only serves the precomputed weights.

File: src/ispc/kernels/lanczos3.ispc [hunk 11]
```diff
@@ -18,97 +18,190 @@ static inline float sinc(float x)
...
+    pixel_ptr[0] = clamp(color[0], 0.0f, 255.0f);
+    pixel_ptr[1] = clamp(color[1], 0.0f, 255.0f);
+    pixel_ptr[2] = clamp(color[2], 0.0f, 255.0f);
+}
+
+uint8<4> sample_4_channels(varying uint64 read_address, const uniform uint8* varying src) {
+    const uniform uint8* pixel_ptr = src + read_address;
+    const uniform uint8<4>* pixel_ptr4 = (const uniform uint8<4>*)(pixel_ptr);
+    varying uint8<4> dst = {0, 0, 0, 0};
+    dst = *pixel_ptr4;
```
Comment (MarijnS95) [inline src/ispc/kernels/lanczos3.ispc:116]: Is the zero-initialization needed?

File: src/ispc/kernels/lanczos3.ispc [hunk 12]
```diff
@@ -18,97 +18,190 @@ static inline float sinc(float x)
...
+    return dst;
+}
+
+void clean_and_write_4_channels(varying float<4> color, varying uint64 write_address, uniform uint8* varying dst) {
```
Comment (MarijnS95) [inline src/ispc/kernels/lanczos3.ispc:120]: Here and everywhere else: the pointer (`* varying`) is `varying` by default (each program instance points to _a different_ pixel in the buffer), and the pointed-to type (`uint8`) is `uniform` by default (all programs point within the same buffer/memory).

File: src/ispc/kernels/lanczos3.ispc [hunk 13]
```diff
@@ -18,97 +18,190 @@ static inline float sinc(float x)
...
-}
 
-export void resample(uniform uint32 width, uniform uint32 height, uniform uint32 stride, uniform uint8 num_channels, uniform uint32 target_width, uniform uint32 target_height, uniform const uint8 src_data[], uniform uint8 out_data[]) {
-    uniform Image src = {src_data, {width, height}};
-    uniform float<2> target_size = {(float)target_width, (float)target_height};
-    uniform float<2> inv_target_size = {1.0 / target_width, 1.0 / target_height};
+    // Accumulate the scratch space data along the height
+    // Downsamples the src_height X target_width image to target_height * target_width
+    foreach_tiled(y =  0 ... target_height, x = 0 ... target_width) {
+
```
Comment (MarijnS95) [inline src/ispc/kernels/lanczos3.ispc:165]:
```suggestion
```

File: src/ispc/kernels/lanczos3.ispc [hunk 14]
```diff
@@ -18,97 +18,190 @@ static inline float sinc(float x)
...
-    return col * 255;
-}
 
-export void resample(uniform uint32 width, uniform uint32 height, uniform uint32 stride, uniform uint8 num_channels, uniform uint32 target_width, uniform uint32 target_height, uniform const uint8 src_data[], uniform uint8 out_data[]) {
-    uniform Image src = {src_data, {width, height}};
-    uniform float<2> target_size = {(float)target_width, (float)target_height};
-    uniform float<2> inv_target_size = {1.0 / target_width, 1.0 / target_height};
+    // Accumulate the scratch space data along the height
+    // Downsamples the src_height X target_width image to target_height * target_width
+    foreach_tiled(y =  0 ... target_height, x = 0 ... target_width) {
```
Comment (MarijnS95) [inline src/ispc/kernels/lanczos3.ispc:164]: Maybe asking a lot, but can we make this entire loop and contents generic over width/height and the buffer that it uses? Effectively making a downsample that only works horizontally or vertically, getting rid of some of the duplication.

File: src/ispc/kernels/lanczos3.ispc [hunk 15]
```diff
@@ -18,97 +18,190 @@ static inline float sinc(float x)
...
 
-    foreach_tiled (y = 0 ... target_height, x = 0 ... target_width) {
-        float yf = y / target_size.y;
-        float xf = x / target_size.x;
+        float<3> color3 = {0.0f, 0.0f, 0.0f};
+        float<4> color4 = {0.0f, 0.0f, 0.0f, 0.0f};
+        for (uint32 i = 0; i < num_vertical_weights; i++) {
+            float weight = vertical_weights[i];
+            uint32 scratch_y = src_height_start + i;
+            uint64 scratch_read_address = (scratch_y * target_width + x) * num_channels;
```
Comment (MarijnS95) [inline src/ispc/kernels/lanczos3.ispc:175]: Are these gathers going to be hard to vectorize? Since this is the innermost loop, I wonder if the loop above could store the pixels transposed (so row-first) so that this increments by `num_channels` instead of by `num_channels * target_width` :)

(IOW: every scratch pixel is only stored once, but read many times)

File: src/ispc/kernels/lanczos3.ispc [hunk 16]
```diff
@@ -18,97 +18,190 @@ static inline float sinc(float x)
...
+            clean_and_write_3_channels(color3, out_write_address, out_data);
+        else
+            clean_and_write_4_channels(color4, out_write_address, out_data);
     }
 }
+
+/// scratch_space must be at least src_height * target_width pixels big
+export void resample_with_cache_3(uniform uint32 src_width, uniform uint32 src_height, uniform uint32 target_width, uniform uint32 target_height,
+    uniform const Cache * uniform cache, uniform uint8 scratch_space[], uniform const uint8 src_data[], uniform uint8 out_data[]) {
+    uniform const uint32 num_channels = 3;
```
Comment (MarijnS95) [inline src/ispc/kernels/lanczos3.ispc:198]: Remind me: did it make a difference to pass this as arugment from Rust? (I hope it does, the compiler should be able to deadstrip the if/else for 3/4 channels after const-folding)

File: src/ispc/mod.rs [hunk 17]
```diff
@@ -1,4 +1,62 @@
 #![allow(deref_nullptr)]
 
 use ispc_rt::ispc_module;
+use std::rc::Rc;
+
+use crate::CachedWeight;
 ispc_module!(downsample_ispc);
+
+// `WeightVariables` is a generated struct, so we cannot realistically add derivable traits to it.
+// Because of this we disable the clippy warning.
```
[the remark this answers is not in the document]
Comment (MarijnS95) [reply src/ispc/mod.rs:10]: https://github.com/Twinklebear/ispc-rs/pull/22 you can now poke at `bindgen::Builder`!

File: src/ispc/mod.rs [hunk 18]
```diff
@@ -1,4 +1,62 @@
...
+
+impl WeightCollection {
+    pub(crate) fn new(mut weights: Vec<CachedWeight>) -> Rc<Self> {
+        let (starts, counts): (Vec<_>, Vec<_>) = weights
+            .iter()
+            .map(|w| (w.start, w.coefficients.len() as u32))
+            .unzip();
+
+        let weights = weights
+            .drain(..)
```
Comment (MarijnS95) [inline src/ispc/mod.rs:40]: Would `into_iter()` (should move the `Vec` and its contents) work here, and allow you to drop the `mut`?

File: src/ispc/mod.rs [hunk 19]
```diff
@@ -1,4 +1,62 @@
...
+
+    // Keep these because we need to keep them in memory
+    _starts: Vec<u32>,
+    _weight_counts: Vec<u32>,
+    _weights: Vec<Rc<Vec<f32>>>,
+    _weights_ptrs: Vec<*const f32>,
+}
+
+impl WeightCollection {
+    pub(crate) fn new(mut weights: Vec<CachedWeight>) -> Rc<Self> {
```
Comment (MarijnS95) [inline src/ispc/mod.rs:33]: Let the caller wrap this in an `Rc` if they so desire, don't force it up-front.

File: src/lib.rs [hunk 20]
```diff
@@ -35,30 +40,163 @@ impl<'a> Image<'a> {
...
+    for v in variables.iter() {
+        let coefficient_count = (v.src_end - v.src_start + 1.0) as u32;
+        // The unique values that define a collection of cached weights are how many pixels it includes and the distance from its start to its center.
+        // We use them to create a key based on which we reuse ones we've calculated previously.
+        let reuse_key = (
+            coefficient_count,
+            (v.src_center - v.src_start).to_ne_bytes(),
+        );
+
+        let reused = reuse_heap.get(&reuse_key);
```
Comment (MarijnS95) [inline src/lib.rs:86]: Nit: there should be a bunch of functions that help with efficiently retrieving or inserting a new item, without doing the lookup multiple times.

File: src/lib.rs [hunk 21]
```diff
@@ -35,30 +40,163 @@ impl<'a> Image<'a> {
...
+
+    res
+}
+
+/// Samples the provided image down to the specified width and height.
+/// `target_width` and `target_height` are expected to be less than or equal to their `src` counter parts.
+/// Will panic if the target dimensions are the same as the source image's.
 ///
-/// Will panic if the target width or height are higher than that of the source image.
+/// For a more fine-tunable version of this function, see [downsample_with_custom_scale].
```
Comment (MarijnS95) [inline src/lib.rs:122]:
```suggestion
/// For a more fine-tunable version of this function, see [`downsample_with_custom_scale()`].
```

File: src/lib.rs [hunk 22]
```diff
@@ -35,30 +40,163 @@ impl<'a> Image<'a> {
...
 pub fn downsample(src: &Image, target_width: u32, target_height: u32) -> Vec<u8> {
+    downsample_with_custom_scale(src, target_width, target_height, 3.0)
+}
+
+/// Version of [downsample] which allows for a custom filter scale, thus trading between speed and final image quality.
```
Comment (MarijnS95) [inline src/lib.rs:127]:
```suggestion
/// Version of [`downsample()`] which allows for a custom filter scale, thus trading between speed and final image quality.
```

File: src/lib.rs [hunk 23]
```diff
@@ -35,30 +40,163 @@ impl<'a> Image<'a> {
...
+///
+/// `filter_scale` controls how many samples are made relative to the size ratio between the source and target resolutions.
+/// The higher the scale, the more detail is preserved, but the slower the downsampling is. Note that the effect on the detail becomes smaller the higher the scale is.
+///
+/// As a guideline, a `filter_scale` of 3.0 preserves detail well.
+/// A scale of 1.0 preserves is good if speed is necessary, but still preserves a decent amount of detail.
```
Comment (MarijnS95) [inline src/lib.rs:133]: preserves _or_ is good?

File: src/lib.rs [hunk 24]
```diff
@@ -35,30 +40,163 @@ impl<'a> Image<'a> {
...
+    } else {
+        WeightCollection::new(calculate_weights(src.height, target_height, filter_scale))
+    };
+
+    // The new implementation needs a src_height * target_width intermediate buffer.
+    let mut scratch_space = Vec::new();
+    scratch_space.resize(
+        (src.height * target_width * src.format.num_channels() as u32) as usize,
+        0u8,
+    );
```
Comment (MarijnS95) [inline src/lib.rs:164]: `vec![...; 0u8]`?

**Comment by MarijnS95** [comment]
Since the gist of this PR is improving quality - and if I understand correctly using weight ""caching"" to get there without outlandish times - how does it fare against `.resize()` from a quality perspective? `main` is already slower than `resize` (45ms vs 37ms) and this PR bumps us to 61ms:

```
Downsample `square_test.png` using ispc_downsampler                                                                            
                        time:   [61.593 ms 61.710 ms 61.838 ms]
                        change: [+36.286% +36.552% +36.809%] (p = 0.00 < 0.05)
                        Performance has regressed.
Found 1 outliers among 100 measurements (1.00%)
  1 (1.00%) high severe

Downsample `square_test.png` using resize                                                                            
                        time:   [37.770 ms 37.803 ms 37.839 ms]
                        change: [-0.2602% -0.1511% -0.0328%] (p = 0.01 < 0.05)
                        Change within noise threshold.
Found 19 outliers among 100 measurements (19.00%)
  7 (7.00%) high mild
  12 (12.00%) high severe
```

File: src/lib.rs [hunk 25]
```diff
@@ -66,28 +73,158 @@ pub fn scale_alpha_to_original_coverage(
...
+// Defines a line of weights. `coefficients` contains a weight for each pixel after `start`
+#[derive(Debug, Clone)]
+struct CachedWeight {
+    pub start: u32,
+    pub coefficients: Rc<Vec<f32>>,
+}
+
+pub(crate) fn calculate_weights(src: u32, target: u32, filter_scale: f32) -> Vec<CachedWeight> {
+    assert!(
+        src > target,
```
Comment (Athosvk) [inline src/lib.rs:85]: We allow resampling in only width or height, but this assertion seems to break this principle
